### ISSUE TITLE
mcstas:component ESS_butterfly Display code makes use of global variables !!

### ISSUE THREAD

**Issue by farhi** [McStasMcXtrace/McCode#631]
The DISPLAY part of the `ESS_butterfly` component includes a:

- %include "ESS_butterfly-geometry.c"

which inserts directly some code depending on internal component variables, assumed global. This should be re-factored as a function call with explicit variables passed, as we have an include from share (must be 'sharable' in some way).

**Comment by farhi** [comment]
Yes, porting to cogen3 is an opportunity to check how code is built. The cogen3 is less tolerant to messy code. I have found many unprotected variables in components (e.g. preventing multiple uses). Also, some tests are not functional anymore.

I will port the butterfly geometry into a function, and call it explicitly in the DISPLAY part of the component.

**Comment by farhi** [comment]
Currently, I have 55% of the tests passing. Working fast on this.

**Comment by willend** [comment]
Sounds good. 

To avoid potential overlaps in effort it could be good if you send a test abstract from time to time? 

Is the "Dresden-cogen3-topdown" the basis of your tests?

**Comment by farhi** [comment]
branch: Dresden-2018-cogen3-openACC-topdown

- Directory: mcstas-comps/examples-full
- Failed:   89
- Compiled: 87

The number of compiled instruments dropped due to a change I made in cogen.c.in, but this is going in the right direction though. Current lists:

Failed:

  ```
    'ESS_2015_test.instr'
    'ESS_butterfly_Guide_curved_test.instr'
    'ESS_butterfly_MCPL_test.instr'
    'ESS_butterfly_tfocus_NOFOCUS_test.instr'
    'ESS_mcpl2hist.instr'
    'FZJ_BenchmarkSfin2.instr'
    'Granroth_SNS_decoupled_poisoned.instr'
    'HZB_FLEX.instr'
    'Histogrammer.instr'
    'ILL_H10_IN8.instr'
    'ILL_H13_IN20.instr'
    'ILL_H142_IN12.instr'
    'ILL_H143_LADI.instr'
    'ILL_H15_IN6.instr'
    'ILL_H16_IN5.instr'
    'ILL_H16_IN5_Mantid.instr'
    'ILL_H22_D1A.instr'
    'ILL_H22_D1B.instr'
    'ILL_H22_VIVALDI.instr'
    'ILL_H25_IN22.instr'
    'ILL_H5.instr'
    'ILL_H53_D16.instr'
    'ILL_H53_IN14.instr'
    'ILL_H5_new.instr'
    'ILL_H8_IN1.instr'
    'ILL_IN13.instr'
    'ILL_IN5.instr'
    'ILL_IN5_Mantid.instr'
    'ILL_IN6.instr'
    'ISIS_CRISP.instr'
    'ISIS_GEM.instr'
    'ISIS_MERLIN.instr'
    'ISIS_OSIRIS.instr'
    'RITA-II.instr'
    'Reflectometer.instr'
    'SAFARI_MPISI.instr'
    'SEMSANS_instrument.instr'
    'SE_example.instr'
    'SE_example2.instr'
    'SNS_ARCS.instr'
    'SNS_test.instr'
    'TestSANS.instr'
    'Test_Fermi.instr'
    'Test_Guides.instr'
    'Test_Guides_Curved.instr'
    'Test_MCPL_input.instr'
    'Test_Magnetic_Constant.instr'
    'Test_Magnetic_Majorana.instr'
    'Test_Magnetic_Rotation.instr'
    'Test_Magnetic_Userdefined.instr'
    'Test_Monochromators.instr'
    'Test_Pol_Bender.instr'
    'Test_Pol_Bender_Vs_Guide_Curved.instr'
    'Test_Pol_Guide_Vmirror.instr'
    'Test_Pol_Mirror.instr'
    'Test_Pol_TripleAxis.instr'
    'Test_PreMonitor_nD.instr'
    'Test_SSR_SSW.instr'
    'Test_SSR_SSW_Guide.instr'
    'Test_Sample_nxs_diffraction.instr'
    'Test_Sample_nxs_imaging.instr'
    'Test_Scatter_log_losses.instr'
    'Test_Scatter_log_mvalues.instr'
    'Test_Scatter_log_ssw_mcnp.instr'
    'Test_Sources.instr'
    'Test_StatisticalChopper.instr'
    'Test_pol_tabled.instr'
    'Test_shellguides.instr'
    'Union_demonstration.instr'
    'Union_demonstration_absorption_image.instr'
    'Union_external_component.instr'
    'Union_external_component_test.instr'
    'Union_incoherent_validation.instr'
    'Union_laue_camera.instr'
    'Union_manual_example.instr'
    'Union_powder_validation.instr'
    'Union_sample_picture_replica.instr'
    'Union_single_crystal_validation.instr'
    'Union_tagging_demo.instr'
    'Union_test_absorption.instr'
    'Union_test_absorption_image.instr'
    'Union_test_box.instr'
    'Union_test_mask.instr'
    'Union_test_powder.instr'
    'Union_time_of_flight.instr'
    'Vin_test.instr'
    'Vout_test.instr'
    'templateSasView.instr'
    'templateSasView_Mantid.instr'
```
working on this...

**Comment by willend** [comment]
Thanks, I've made a dedicated issue for tracking this: https://github.com/McStasMcXtrace/McCode/issues/632

**Comment by farhi** [comment]
code ready to commit for this issue (ESS_butterfly). Will commit most probably tomorrow or Thursday.

**Comment by farhi** [comment]
Fixed with 66e70959bd60e5c80e8f21a4b29cbdb441f528cc
